### PR TITLE
Upsample before crop in `process_mask` to prevent mask leakage

### LINKED ISSUE #24272: Segmentation edges are not smooth
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

I have found that the YOLO instance segmentation model (e.g., yolov8n-seg.pt and other instance segmentation models) may produce segmentation contours that protrude beyond the bounding box. I have localized the issue to the process_mask post-processing function.

Image

Here is the original `ultralytics.utils.ops.process_mask` post-processing function:
```
 def process_mask(protos, masks_in, bboxes, shape, upsample: bool = False):
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # NHW

    width_ratio = mw / shape[1]
    height_ratio = mh / shape[0]
    ratios = torch.tensor([[width_ratio, height_ratio, width_ratio, height_ratio]], device=bboxes.device)

    masks = crop_mask(masks, boxes=bboxes * ratios)  # NHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode="bilinear")[0]  # NHW
    return masks.gt_(0.0).byte()
```
The issue arises because the original algorithm first applies crop_mask and then upsamples (which may interpolate other values), leading to binarization and causing segmentation contours to possibly extend beyond the bounding box.

The improved method is to upsample first and then apply crop_mask:
```
def process_mask(protos, masks_in, bboxes, shape, upsample: bool = False):
    c, mh, mw = protos.shape  # CHW
    masks = (masks_in @ protos.float().view(c, -1)).view(-1, mh, mw)  # NHW

    width_ratio = mw / shape[1]
    height_ratio = mh / shape[0]
    ratios = torch.tensor([[width_ratio, height_ratio, width_ratio, height_ratio]], device=bboxes.device)

    # masks = crop_mask(masks, boxes=bboxes * ratios)  # NHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode="bilinear")[0]  # NHW
    masks = crop_mask(masks, boxes=bboxes)  # 
    return masks.gt_(0.0).byte()
```
Image

### Environment

Ultralytics 8.4.23 🚀 Python-3.10.0 torch-2.10.0+cu128 CUDA:0 (NVIDIA GeForce RTX 3070, 7967MiB)
Setup complete ✅ (16 CPUs, 31.2 GB RAM, 96.7/373.8 GB disk)

OS                     Linux-6.12.63-amd64-desktop-rolling-x86_64-with-glibc2.38
Environment            Linux
Python                 3.10.0
Install                pip
Path                   /home/PKing/miniconda3/envs/py310/lib/python3.10/site-packages/ultralytics
RAM                    31.16 GB
Disk                   96.7/373.8 GB
CPU                    Intel Core(TM) i7-10700K 3.80GHz
CPU count              16
GPU                    NVIDIA GeForce RTX 3070, 7967MiB
GPU count              1
CUDA                   12.8

numpy                  ✅ 1.24.3>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.13.0.92>=4.6.0
pillow                 ✅ 12.1.1>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.10.0>=1.8.0
torch                  ✅ 2.10.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.25.0>=0.9.0
psutil                 ✅ 7.1.1>=5.8.0
polars                 ✅ 1.39.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
    model_file = "yolov8n-seg.pt"
    model = YOLO(model_file)
    results = model.predict('bus.jpg', imgsz=640, save=True)  
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by wjddnwp29**
## Summary
`process_mask(upsample=True)` currently crops at low resolution and then bilinearly upsamples, so the hard `0 → positive` boundary produced by `crop_mask` gets smeared during interpolation, leaving binarized mask pixels outside the image-space bounding box. Reordering the two steps to **upsample → crop**, matching the existing `process_mask_native` behavior, eliminates the leakage. The `upsample=False` path is preserved unchanged for backward compatibility.

Closes #24272.

## Before / after
With a deterministic input (uniformly positive 32×40×40 prototypes, bbox `[40, 40, 120, 120]`, output shape `(160, 160)`):

| | pixels set outside bbox (should be 0) |
|---|---|
| Before | **656** |
| After  | **0** |

`process_mask_native` already produced 0 with the same input, confirming the fix only aligns `process_mask` with the existing correct behavior.

> I will attach a side-by-side visualization (`compare_24272.png`) in a follow-up comment.

## Change
`ultralytics/utils/ops.py` — `process_mask`:
- `upsample=True` path: `F.interpolate(..., bilinear)` first, then `crop_mask` at image resolution (matches `process_mask_native`).
- `upsample=False` path: unchanged (crop at mask resolution with scaled bbox).

## Test
Added `test_process_mask_crops_within_bbox` to `tests/test_python.py` (next to `test_utils_ops`). It asserts no mask pixels exist above/below/left/right of the bbox after `process_mask(..., upsample=True)`.

Local run:
```
tests/test_python.py::test_process_mask_crops_within_bbox PASSED  [100%]
```

`ruff format --check` and `ruff check` both pass on the modified files.

## Notes
- Total effective change: ~10 lines in `ops.py`, ~18 lines in `tests/test_python.py` (29 insertions / 5 deletions).
- No public API change; signature and return shape of `process_mask` are unchanged.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes mask leakage in `process_mask(..., upsample=True)` by changing the operation order to upsample before cropping and adding a regression test for issue #24272. 🎯

### 📊 Key Changes
- Updated `ultralytics/utils/ops.py` so `process_mask()` now performs bilinear upsampling before `crop_mask()` when `upsample=True`.
- Preserved the existing low-resolution crop path for `upsample=False`, including bbox scaling to prototype-space coordinates.
- Added a regression test in `tests/test_python.py` to verify masks stay strictly within the bounding box after upsampling. 🧪
- Documented the root cause in code comments: cropping before bilinear upsampling could smear mask boundaries and create pixels outside the bbox. 🛠️

### 🎯 Purpose & Impact
- Resolves a segmentation bug where masks could extend beyond their intended image-space bounding boxes.
- Improves reliability of postprocessing for detection and segmentation workflows using `process_mask(..., upsample=True)`.
- Reduces the risk of incorrect mask visualization or downstream bbox-constrained mask usage.
- Strengthens long-term stability with a focused regression test to prevent future reintroduction of the bug. ✅

---

## Maintainer update

Rebased onto current `main` (keeps the `masks_in.shape[0] == 0` guard from #25032), trimmed the comment, and removed the bundled regression test in favor of empirical verification.

Deleted: the unconditional `crop_mask(masks, boxes=bboxes * ratios)` call before the upsample branch (relocated into the two resolution-specific branches) and the bundled regression test. Net +7/−5 because the crop cannot be deleted, only relocated after upsampling, and the two branches return different resolutions so they cannot share one crop call.

Verified empirically (32×40×40 all-ones protos, bbox (40,40,120,120), 160×160 image): `main` leaks mask pixels outside the bbox after upsampling; this branch leaks 0 with the 6400-pixel interior fully preserved. The `upsample=False` path is byte-identical to `main`, and the empty-batch path still returns `(0, H, W)` uint8.

### CODE REVIEW

**Comment by wjddnwp29** [comment]
I have read the CLA Document and I sign the CLA

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wjddnwp29, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you shortly. Please review the checklist below to help ensure a smooth merge:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. Your added regression test for #24272 is appreciated 🧪
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The `process_mask()` change looks aligned with the reported leakage bug, and the regression coverage is helpful. One gap: the new test only checks for absence of leakage, so it would miss failures that zero the mask entirely. Adding an interior-positive assertion would make the regression test much stronger.

💬 Posted 1 inline comment

File: tests/test_python.py [hunk 1]
```diff
@@ -579,6 +579,26 @@ def test_utils_ops():
...
+    # which the pre-fix bilinear upsample then smeared into pixels outside the image-space bbox.
+    protos = torch.ones(32, 40, 40)
+    masks_in = torch.ones(1, 32)
+    bboxes = torch.tensor([[40.0, 40.0, 120.0, 120.0]])
+    shape = (160, 160)
+
+    masks = process_mask(protos, masks_in, bboxes, shape, upsample=True)
+    assert masks.shape == (1, 160, 160)
+    m = masks[0]
+    assert m[:40].sum().item() == 0, "mask leaked above the bbox"
```
Comment (UltralyticsAssistant) [inline tests/test_python.py]: 📝 **LOW**: This regression test can still pass if `process_mask()` returns an all-zero mask, because it only asserts that pixels outside the bbox are zero. Since this change also switches the crop to image-space coordinates, please add an assertion that some pixels inside `[40:120, 40:120]` remain set so the test actually catches a broken upsample/crop path.

Comment (wjddnwp29) [reply tests/test_python.py]: Thanks for the catch. Updated the test in 09bf030 to assert `m[40:120, 40:120].sum() == 80*80` (rules out an all-zero mask) and to also exercise the `upsample=False` branch with the analogous containment check at prototype scale.

**Comment by wjddnwp29** [comment]
Pushed 09bf030 addressing the inline review feedback:

- `tests/test_python.py::test_process_mask_crops_within_bbox` now also asserts the bbox interior is fully set (catches all-zero / broken-mask regressions).
- Added the same containment check for the `upsample=False` path so both branches of `process_mask` are exercised.

**Comment by wjddnwp29** [comment]
Apologies — I accidentally clicked the close button while looking for "Resolve conversation" on the inline review thread. Reopened immediately; no changes were intended. The PR is otherwise unchanged.

**Comment by glenn-jocher** [comment]
No problem, thanks for clarifying and for updating the tests to cover both `process_mask` branches; the PR state looks fine now and we’ll continue the review from here.

**Comment by wjddnwp29** [comment]
Hi @onuralpszr @glenn-jocher — noticed this was moved to draft last week. I've since rebased onto latest main and all CI checks are passing. Happy to address any additional feedback (e.g., other callers of process_mask, edge cases, or further test coverage). Let me know what would help move this forward. Thanks!

**Comment by glenn-jocher** [comment]
Thanks for the follow-up and for rebasing; nothing additional is needed from your side right now, so please leave the PR as-is for the next maintainer review pass, and we’ll comment here if any further changes are needed.

**Review by glenn-jocher — APPROVED** [review]
LGTM at exact head 43fffdc. The change relocates cropping to the correct post-upsample coordinate space while preserving the existing non-upsample and empty-batch paths. Delete/replace review: no additional code can be deleted without losing either path; the stale test hunk was removed. Trigger review: no condition suppresses the leakage symptom—the operation itself is moved to the correct owner/order.
